metrics/time/launch_times.sh

```diff
@@ -0,0 +1,172 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#  Description of the test:
+#  This test takes a number of time measurements through the complete
+#  launch/shutdown cycle of a single container.
+#  From those measurements it derives a number of time measures, such as:
+#   - time to payload execution
+#   - time to get to VM kernel
+#   - time in VM kernel boot
+#   - time to quit
+#   - total time (from launch to finished)
+#
+# Note, the <image> used for this test must support the full 'date' command
+# syntax - the date from busybox for isntance *does not* support this, so
+# will not work with this test.
+#
+# Note, this test launches a single container at a time, that quits - thus,
+# this test measures times for the 'first container' only. This test does
+# not look for any scalability slowdowns as the number of running containers
+# increases for instance - that is handled in other tests
+
+set -e
+
+if [ $# -ne 3 ] && [ $# -ne 4 ]; then
+	echo >&2 "Usage: $0 <image> <runtime> <times to run> (<scaling>)"
+	exit 1
+fi
+
+SCRIPT_PATH=$(dirname "$(readlink -f "$0")")
+source "${SCRIPT_PATH}/../lib/common.bash"
+
+IMAGE="$1"
+RUNTIME="$2"
+TIMES="$3"
+SCALING="$4"
+
+# Calculating the kernel time from dmesg stamps only really works for VM
+# based runtimes - we dynamically enable it if we find we are using a known
+# VM runtime
+CALCULATE_KERNEL=
+
+# We set the generic name here, but we save the different time results separately,
+# and append the actual detail to the name at the time of saving...
+TEST_NAME="boot times"
+# If we are scaling, note that in the name
+if [ ! -z "$SCALING" ]; then
+	TEST_NAME="${TEST_NAME} scaling"
+fi
+
+TEST_ARGS="image=${IMAGE} runtime=${RUNTIME} units=seconds"
+
+REQUIRED_CMDS=("bc" "awk")
+
+# The date command format we use to ensure we capture the ns timings
+# Note the no-0-padding - 0 padding the results breaks bc in some cases
+DATECMD="date -u +%-s:%-N"
+
+
+# convert a 'seconds:nanoseconds' string into nanoseconds
+sn_to_ns() {
+	seconds=$(echo $1 | awk -F ':' '{print $1}')
+	ns=$(echo $1 | awk -F ':' '{print $2}')
+
+	totalns=$(( (seconds * 1000000000) + $ns))
+
+	echo $totalns
+}
+
+# convert 'nanoseconds' (since epoch) into a 'float' seconds
+ns_to_s() {
+	echo $(bc <<< "scale=3; $1 / 1000000000")
+}
+
+# Ensure we have yanked down the necessary images before we begin - we do not
+# want to be adding that overhead into any of our measurement times
+function prewarm(){
+	echo Pre-warming by pulling $IMAGE
```

Comment:
missing quotes (") ?

Reply:
well, technically not necessary I believe - echo will echo all of the non-param args I believe - but, sure, for consistency, I'll add ;-)